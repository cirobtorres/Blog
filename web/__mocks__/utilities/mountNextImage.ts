const nextOptimizedImage = (filePath: string) => {
  if (filePath) {
    const filename = filePath.substring(
      filePath.lastIndexOf("/") + 1,
      filePath.lastIndexOf(".")
    );
    const extension = filePath.substring(filePath.lastIndexOf(".") + 1);
    const path = filePath.substring(0, filePath.lastIndexOf("/") + 1);
    const URL = `${path}${filename}.${extension}`;
    const encodedURL = encodeURIComponent(URL);
    const width = "w=3840";
    const quality = "q=75";
    const nextOptimizedImage = `/_next/image?url=${encodedURL}&${width}&${quality}`;
    return nextOptimizedImage;
  }
  return null;
};

interface ToEncodeURIAttributes {
  imageFilename: string;
  imageFileExtension: string;
  path: string;
  width: number;
}

function mountLocalImage({
  imageFilename,
  imageFileExtension,
  path,
  width,
}: ToEncodeURIAttributes) {
  // Mock local images
  const quality = 75;
  const URL = `${path}${imageFilename}.${imageFileExtension}`;
  const encodedURL = encodeURIComponent(URL);
  const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${width}&q=${quality}`;
  return nextOptimizedImage;
}

function requestBackEndImage(
  filePath?: string,
  width: number = 1920,
  options?: {
    PROTOCOL?: string;
    IP?: string;
    PORT?: string;
  }
) {
  // Mock images managed by Strapi
  const quality = 75;

  if (filePath) {
    const filename = filePath.substring(
      filePath.lastIndexOf("/") + 1,
      filePath.lastIndexOf(".")
    );
    const extension = filePath.substring(filePath.lastIndexOf(".") + 1);
    const path = filePath.substring(0, filePath.lastIndexOf("/") + 1);

    const protocol = options ? options.PROTOCOL : "http";
    const ip = options ? options.IP : "127.0.0.1";
    const port = options ? options.PORT : "1337";
    const backendURL = `${protocol}://${ip}:${port}`;

    const URL = `${backendURL}${path}${filename}.${extension}`;

    const encodedURL = encodeURIComponent(URL);
    const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${
      width * 2
    }&q=${quality}`;

    return nextOptimizedImage;
  }

  const URL = "/images/not-authenticated.png";
  const encodedURL = encodeURIComponent(URL);

  const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${
    width * 2
  }&q=${quality}`;

  return nextOptimizedImage;
}

export { nextOptimizedImage, mountLocalImage, requestBackEndImage };
