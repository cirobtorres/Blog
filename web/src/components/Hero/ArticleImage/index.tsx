"use server";

import Image from "next/image";

const ArticleImage = ({ cover }: { cover: Cover }) => {
  return (
    <section
      className={
        "grid grid-cols-article" +
        " max-lg:grid-cols-article-1024" +
        " max-w-screen-2xl mx-auto mb-20"
      }
    >
      <figure
        className={
          "flex flex-col h-[25rem] col-start-2 max-[800px]:col-start-1 max-[800px]:col-span-2 max-lg:mx-4 max-lg:pl-4 max-[800px]:pl-0"
        }
      >
        <div className="shrink-0 relative h-full">
          <Image
            src={`http://127.0.0.1:1337${cover.url}`}
            alt={cover.alternativeText}
            fill
            sizes={
              `(max-width: ${cover.width}) 100vw` +
              `, (max-width: ${cover.width / 2}) 50vw`
              // `, 33vw`
            }
            className="absolute object-cover"
          />
        </div>
        {cover.caption && (
          <figcaption className="mx-2 mt-3 text-xs">{cover.caption}</figcaption>
        )}
      </figure>
    </section>
  );
};

export default ArticleImage;
