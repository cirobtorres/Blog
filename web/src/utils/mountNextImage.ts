interface ToEncodeURIAttributes {
  imageFilename: string;
  imageFileExtension: string;
  path: string;
  width: number;
  quality: number;
}

function mountNextImage({
  imageFilename,
  imageFileExtension,
  path,
  width,
  quality,
}: ToEncodeURIAttributes) {
  const URL = `${path}${imageFilename}.${imageFileExtension}`;
  const encodedURL = encodeURIComponent(URL);
  const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${width}&q=${quality}`;
  return nextOptimizedImage;
}

function requestBackEndNextImage(
  imageFilename: string,
  imageFileExtension: string,
  path: string,
  width: number,
  quality: number,
  options?: {
    PROTOCOL?: string;
    IP?: string;
    PORT?: string;
  }
) {
  const URL = `${options ? options.PROTOCOL : "http"}://${
    options ? options.IP : "127.0.0.1"
  }:${
    options ? options.PORT : "1337"
  }${path}${imageFilename}.${imageFileExtension}`;
  const encodedURL = encodeURIComponent(URL);
  const nextOptimizedImage = `/_next/image?url=${encodedURL}&w=${width}&q=${quality}`;

  return nextOptimizedImage;
}

export { mountNextImage, requestBackEndNextImage };
