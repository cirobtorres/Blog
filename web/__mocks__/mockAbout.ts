import { faker } from "@faker-js/faker";
import { randomInt } from "./utilities/randomInt";
import { generateRandomCreatedAt } from "./utilities/randomISOdatetime";

export const createAboutMock = () => ({
  documentId: faker.string.uuid(),
  title: faker.lorem.words(randomInt(5, 12)),
  ...(() => {
    const github_link = faker.internet.url();
    const github_blog_link = github_link + "/" + faker.lorem.slug(3);
    return { github_link, github_blog_link };
  })(),
  ...generateRandomCreatedAt(),
  blocks: [],
});
