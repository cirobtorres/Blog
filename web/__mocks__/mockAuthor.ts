import { faker } from "@faker-js/faker";
import { randomInt } from "./utilities/randomInt";
import { randomRelativeURL } from "./utilities/randomRelativeURL";

export const createAuthorMock = (avatar = true) => ({
  documentId: faker.string.uuid(),
  name: faker.person.firstName(),
  avatar: avatar
    ? {
        documentId: faker.string.uuid(),
        ...randomRelativeURL(),
        alternativeText: faker.lorem.sentence(randomInt(5, 10)),
      }
    : null,
});
