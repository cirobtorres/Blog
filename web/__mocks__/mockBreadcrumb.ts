import { faker } from "@faker-js/faker";
import { randomInt } from "./utilities/randomInt";
import slugify from "@/utils/slugify";

export const createBreadcrumbMock = () => ({
  documentId: faker.string.uuid(),
  ...(() => {
    const name = faker.lorem.words(randomInt(2, 5));
    const slug = slugify(name);
    return { name, slug };
  })(),
  description: faker.lorem.sentence(randomInt(5, 12)),
});
