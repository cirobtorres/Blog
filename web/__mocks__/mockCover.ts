import { faker } from "@faker-js/faker";
import { randomInt } from "./utilities/randomInt";
import { randomRelativeURL } from "./utilities/randomRelativeURL";

export const createCoverMock = (width = 512, height = 512) => ({
  documentId: faker.string.uuid(),
  ...randomRelativeURL(),
  alternativeText: faker.lorem.sentence(randomInt(5, 10)),
  caption: faker.lorem.sentence(randomInt(10, 20)),
  width,
  height,
});
