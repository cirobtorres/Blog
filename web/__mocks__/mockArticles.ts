import { faker } from "@faker-js/faker";
import { randomInt } from "./utilities/randomInt";
import { generateRandomCreatedAt } from "./utilities/randomISOdatetime";
import slugify from "@/utils/slugify";
import { createCoverMock } from "./mockCover";
import { createAuthorMock } from "./mockAuthor";
import { mockTopic } from "./mockTopic";

export function createArticleMock() {
  return {
    documentId: faker.string.uuid(),
    ...(() => {
      const title = faker.lorem.sentence(randomInt(5, 12));
      const slug = slugify(title);
      return { title, slug };
    })(),
    description: faker.lorem.sentence(randomInt(14, 28)),
    ...generateRandomCreatedAt(),
    cover: createCoverMock(),
    author: createAuthorMock(),
    topic: mockTopic(),
    tools: [],
    tags: [],
    blocks: [],
  };
}

export const mockArticles = (length = randomInt(5, 10)) =>
  Array.from({ length }, () => ({
    ...createArticleMock(),
  }));

export const mockArticle = createArticleMock();

export const mockArticlesCount = {
  articles_connection: {
    pageInfo: { total: 100, pageCount: 10, pageSize: 10, page: 1 },
  },
};
