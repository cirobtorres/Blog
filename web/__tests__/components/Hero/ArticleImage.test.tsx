import ArticleImage from "@/components/Hero/ArticleImage";
import { requestBackEndImage } from "../../../__mocks__/utilities/mountNextImage";
import { render, screen } from "@testing-library/react";
import { createCoverMock } from "../../../__mocks__/mockCover";
import { faker } from "@faker-js/faker";

faker.seed(1); // Snapshot

const mockCover = createCoverMock(512, 512);

describe("ArticleImage", () => {
  it("renders article image component", () => {
    render(<ArticleImage cover={mockCover} />);
    const articleImage = screen.getByTestId("hero-article-image");
    expect(articleImage).toBeInTheDocument();
  });

  it("renders article image", () => {
    render(<ArticleImage cover={mockCover} />);
    const image = screen.getByTestId("hero-article-image-cover");
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute("src", requestBackEndImage(mockCover.url));
  });

  it("renders article image alt text", () => {
    render(<ArticleImage cover={mockCover} />);
    const image = screen.getByTestId("hero-article-image-cover");
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute("role", "img");
    expect(image).toHaveAttribute("alt", mockCover.alternativeText);
  });

  it("renders article image alt text", () => {
    render(<ArticleImage cover={mockCover} />);
    const caption = screen.getByTestId("hero-article-image-caption");
    expect(caption).toBeInTheDocument();
    expect(caption).toHaveAttribute("role", "doc-figure");
    expect(caption).toHaveTextContent(mockCover.caption);
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(<ArticleImage cover={mockCover} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
