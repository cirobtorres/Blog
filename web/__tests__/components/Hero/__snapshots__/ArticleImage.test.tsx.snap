// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleImage matches the snapshot 1`] = `
<DocumentFragment>
  <section
    class="grid grid-cols-article max-lg:grid-cols-article-1024 max-w-screen-2xl mx-auto mb-20"
    data-testid="hero-article-image"
  >
    <figure
      aria-labelledby="image-caption-6b042125-686a-43e0-8a68-23cf5bee102e"
      class="flex flex-col h-[25rem] col-start-2 max-[800px]:col-start-1 max-[800px]:col-span-2 max-lg:mx-4 max-lg:pl-4 max-[800px]:pl-0"
    >
      <div
        class="shrink-0 relative h-full"
      >
        <img
          alt="Occaecati venio apto apud timor cubicularis asperiores vestigium conqueror tantillus."
          class="absolute object-cover"
          data-nimg="fill"
          data-testid="hero-article-image-cover"
          decoding="async"
          loading="lazy"
          role="img"
          sizes="(max-width: 512) 100vw, (max-width: 256) 50vw, 33vw"
          src="/_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fagitated%2Fhyena%2Ffor%2Fha%2Fpsst.png&w=3840&q=75"
          srcset="/_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fagitated%2Fhyena%2Ffor%2Fha%2Fpsst.png&w=256&q=75 256w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fagitated%2Fhyena%2Ffor%2Fha%2Fpsst.png&w=384&q=75 384w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fagitated%2Fhyena%2Ffor%2Fha%2Fpsst.png&w=640&q=75 640w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fagitated%2Fhyena%2Ffor%2Fha%2Fpsst.png&w=750&q=75 750w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fagitated%2Fhyena%2Ffor%2Fha%2Fpsst.png&w=828&q=75 828w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fagitated%2Fhyena%2Ffor%2Fha%2Fpsst.png&w=1080&q=75 1080w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fagitated%2Fhyena%2Ffor%2Fha%2Fpsst.png&w=1200&q=75 1200w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fagitated%2Fhyena%2Ffor%2Fha%2Fpsst.png&w=1920&q=75 1920w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fagitated%2Fhyena%2Ffor%2Fha%2Fpsst.png&w=2048&q=75 2048w, /_next/image?url=http%3A%2F%2F127.0.0.1%3A1337%2Fagitated%2Fhyena%2Ffor%2Fha%2Fpsst.png&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
      <figcaption
        class="mx-2 mt-3 text-xs"
        data-testid="hero-article-image-caption"
        role="doc-figure"
      >
        Vacuus quia tantillus conscendo centum vehemens cursus vobis somnus quasi amplus vir defaeco nesciunt cumque capillus venio.
      </figcaption>
    </figure>
  </section>
</DocumentFragment>
`;
